### PR TITLE
Fix password displayed as username in Firefox password manager dialog

### PR DESCRIPTION
**Pull request by danxuliu**
Fixes (in master) #7031

The problem was (besides the [password manager ignoring `autocomplete="off"`](https://developer.mozilla.org/en-US/docs/Web/Security/Securing_your_site/Turning_off_form_autocompletion#The_autocomplete_attribute_and_login_fields)) that the password input element (in the HTML form) is cloned in a text input element, which is used to show the password in plain text when clicking on the _Show password_ button. As it was a text input immediately followed by a password input Firefox seemed to assume that it had to be the username and ignored the real username field, no matter the value set for the `autocomplete` attribute (`autocomplete="off"` for the whole form, `autocomplete="street-name"` (because it is a [supported `autocomplete` value](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#attr-autocomplete) that has nothing to do with login) or `autocomplete="new-password"` for the cloned text input, `autocomplete="name/username/email"` for the username text input, nor combinations of those).

Now the cloned text input is added after the password input, so Firefox no longer thinks that the cloned text input is the username field and the password manager dialog shows the proper username instead.

Chromium was not affected by the original issue (it also offers to save the credentials, but its password manager dialog showed the proper username), and it still works as before after the change.

### CODE REVIEW

**Review by rullzer — APPROVED** [review]
Sure :)

**Comment by MorrisJobke** [comment]
* [ ] the password strength meter on the personal settings page now moves above the input field:

bildschirmfoto 2018-01-05 um 10 47 36
bildschirmfoto 2018-01-05 um 10 47 41

**Comment by danxuliu** [comment]
>  the password strength meter on the personal settings page now moves above the input field

Thanks, I had not noticed it. The problem is that _strengthify_ also [adds the password strength meter after the element it is applied on](https://github.com/nextcloud/server/blob/afcb41886ecaf3b05a21f4379f6386fba5ff7954/core/vendor/strengthify/jquery.strengthify.js#L179) and as [it is applied on `#pass2`](https://github.com/nextcloud/server/blob/4c3eb80a9102ceaf08d5143eaba8877a8a43e57d/settings/js/settings/personalInfo.js#L388) after [calling showPassword](https://github.com/nextcloud/server/blob/4c3eb80a9102ceaf08d5143eaba8877a8a43e57d/settings/js/settings/personalInfo.js#L144) it ends between the password input and the plain text input clone.

As the text input clone is updated when the password input is updated and vice versa I will check if it is possible to apply _strengthify_ to the text input clone instead.

**Comment by danxuliu** [comment]
>  I will check if it is possible to apply strengthify to the text input clone instead.

No, that approach is not valid due to how the events are triggered in the password input and the text input clone.

The strength bar problem can be fixed (I have tested it), besides using absolute CSS positions (which I would prefer to avoid), by making possible to specify the element that the strength bar should be added after.

@MorrisJobke I guess that I should send a pull request with the needed change in https://github.com/MorrisJobke/strengthify, and then just update the version in [bower.json](https://github.com/nextcloud/server/blob/6d2bff46c95615a774e9701c02da0cd4fc3561b5/bower.json#L28) and the code in _core/vendor_, right? Am I missing anything?

**Comment by MorrisJobke** [comment]
That’s right - maybe we should not use it as external dependency and just maintain it in here as I don’t maintain it anyways - shame on me :/

**Comment by rullzer** [comment]
So what is the status here? Fix soonish? Move to 14?

**Comment by danxuliu** [comment]
@rullzer
>  So what is the status here? Fix soonish? Move to 14?

It is almost done; needs https://github.com/MorrisJobke/strengthify/pull/20 and using the new parameter when calling `strengthify` here.

**Comment by MorrisJobke** [comment]
> It is almost done; needs MorrisJobke/strengthify#20 and using the new parameter when calling strengthify here.

Just released it as 0.5.3: https://github.com/MorrisJobke/strengthify/releases/tag/0.5.3

**Comment by rullzer** [comment]
Bumd the js dep... but it already looks ok now in chromium for me...

**Comment by rullzer** [comment]
ah no caching :P

**Comment by danxuliu** [comment]
Thanks for the library bump :-)

I have pushed the commit that uses the new `$addAfter` and [fixes the strength bar changing its position when the password is shown](https://github.com/nextcloud/server/pull/7709#issuecomment-355514262).

It should fix it too in the _Renew password_ of LDAP... although **I have not tested** it myself in that case, just updated the code.

No changes were needed in the installation page, as in that case `showPassword` was called after `strengthify`.

**Review by rullzer — APPROVED** [review]
whoohoo! still works!

File: apps/user_ldap/js/renewPassword.js [hunk 1]
```diff
@@ -46,5 +46,6 @@ $(document).ready(function() {
 			t('core', 'Strong password')
 		],
 		drawTitles: true,
+		$addAfter: $('input[name="newPassword-clone"'),
```
Comment (MorrisJobke) [inline apps/user_ldap/js/renewPassword.js]: There is a missing `]` which is also logged in the JS console. Let me fix this.

File: settings/js/settings/personalInfo.js [hunk 2]
```diff
@@ -395,6 +395,7 @@ $(document).ready(function () {
 			t('settings', 'Strong password')
 		],
 		drawTitles: true,
+		$addAfter: $('input[name="newpassword-clone"'),
```
Comment (MorrisJobke) [inline settings/js/settings/personalInfo.js]: Same here 🙈

Comment (MorrisJobke) [reply settings/js/settings/personalInfo.js]: bildschirmfoto 2018-01-11 um 14 07 27

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by danxuliu** [comment]
Backport is in #7797

**Pull request merged**